File: backend/src/main/java/com/plugga/backend/controller/PileController.java
package com.plugga.backend.controller;

import com.plugga.backend.entity.Pile;
import com.plugga.backend.service.PileService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/piles")
public class PileController {
    private PileService pileService;

    @Autowired
    public PileController(PileService pileService){
        this.pileService = pileService;
    }

    @GetMapping("/")
    public List<Pile> findAll(){
        return pileService.findAll();
    }

    @GetMapping("/{pileId}")
    public Pile getPile(@PathVariable int pileId){
        Pile pile = pileService.findById(pileId);
        if (pile == null){
            throw new RuntimeException("Could not find pile using id: " + pileId);
        }
        return pile;
    }

    @PostMapping("/")
    public Pile addPile(@RequestBody Pile pile){
        pile.setId(0);
        pileService.save(pile);
        return pile;
    }

    @PutMapping("/")
    public Pile updatePile(@RequestBody Pile pile){
        pileService.save(pile);
        return pile;
    }

    @DeleteMapping("/{pileId}")
    public String deletePile(@PathVariable int pileId){
        Pile tempPile = pileService.findById(pileId);
        if (tempPile == null){
            throw new RuntimeException("Could not find pile using id: " + pileId);
        }
        pileService.deleteById(pileId);
        return "Deleted pile with id: " + pileId;
    }
}

File: backend/src/main/kotlin/com/plugga/backend/controller/PileController.kt
package com.plugga.backend.controller

import com.plugga.backend.entity.Pile
import com.plugga.backend.service.PileService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PagedResourcesAssembler
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.PagedModel
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/piles")
class PileController
@Autowired
constructor(private val pileService: PileService) {

    @GetMapping("")
    fun findAll(
        pageable: Pageable,
        pagedResourcesAssembler: PagedResourcesAssembler<Pile>
    ): PagedModel<EntityModel<Pile>> {
        return pagedResourcesAssembler.toModel(pileService.findAll(pageable))
    }

    @GetMapping("/{pileId}")
    fun getPile(@PathVariable pileId: Int): Pile {
        return pileService.findById(pileId) ?: throw RuntimeException("Could not find pile using id: $pileId")
    }

    @GetMapping(value = [""], params = ["deckId"])
    fun getByDeckId(
        pageable: Pageable,
        pagedResourcesAssembler: PagedResourcesAssembler<Pile>,
        @RequestParam("deckId") deckId: Int
    ): PagedModel<EntityModel<Pile>> {
        return pagedResourcesAssembler.toModel(pileService.findByDeckId(pageable, deckId))
    }

    @PostMapping("")
    fun addPile(@RequestBody pile: Pile): Pile {
        pile.id = 0
        pileService.save(pile)
        return pile
    }

    @PutMapping("/{pileId}")
    fun updatePile(@PathVariable pileId: Int, @RequestBody pile: Pile): Pile {
        pile.id = pileId
        pileService.save(pile)
        return pile
    }

    @DeleteMapping("/{pileId}")
    fun deletePile(@PathVariable pileId: Int): String {
        pileService.findById(pileId) ?: throw RuntimeException("Could not find pile using id: $pileId")
        pileService.deleteById(pileId)
        return "Deleted pile with id: $pileId"
    }
}
